src/main.rs

```diff
@@ -495,3 +512,38 @@ fn main_recv(addr: &str, n_conn: &str) -> Result<()> {
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{thread, time::Duration};
+
+    #[test]
+    fn test_accepts_valid_protocol() {
+        thread::spawn(|| {
+            std::fs::File::create("a-file").unwrap();
+            main_send("127.0.0.1:1234", &["a-file".into()], 1).unwrap();
+        });
+        thread::sleep(Duration::from_millis(100));
+
+        main_recv("127.0.0.1:1234", "1", false, 1).unwrap();
+    }
+
+    #[test]
+    fn test_refuses_invalid_protocol() {
+        thread::spawn(|| {
+            std::fs::File::create("a-file").unwrap();
+            main_send("127.0.0.1:3333", &["a-file".into()], 2).unwrap();
+        });
+        thread::sleep(Duration::from_millis(100));
+
+        let res = main_recv("127.0.0.1:3333", "1", false, 1);
+        match res {
+            Ok(()) => panic!("Expected failure"),
+            Err(e) => match e.kind() {
+                ErrorKind::InvalidData => (),
+                other => panic!("Expected InvalidData, got {other}"),
+            },
+        }
```

Comment:
`panic!` on `Ok` is also called `.err().unwrap()` (for the default message) or `.err().expect("custom message")`. `ErrorKind` implements `Eq` and `Debug`, so we can even go a step further and use `assert_eq!`:

```suggestion
        assert_eq!(res.err().expect("Expected failure").kind(), ErrorKind::InvalidData);
```